src/Xrefcheck/Config.hs

```diff
@@ -57,18 +65,119 @@ makeLensesWith postfixFields ''VerifyConfig
 -- Default config
 -----------------------------------------------------------
 
+defConfigUnfilled :: ByteString
+defConfigUnfilled =
+  $(embedFile ("src-files" </> "def-config.yaml"))
+
+-- | Picks raw config with @:PLACEHOLDER:<key>:@ and fills the specified fields
+-- in it, picking a replacement suitable for the given key. Only strings and lists
+-- of strings can be filled this way.
+--
+-- This will fail if any placeholder is left unreplaced, however extra keys in
+-- the provided replacement won't cause any warnings or failures.
+fillHoles
+  :: HasCallStack
+  => [(ByteString, Either ByteString [ByteString])] -> ByteString -> ByteString
+fillHoles allReplacements rawConfig =
+  let holesLocs = R.getAllMatches $ holeLineRegex `R.match` rawConfig
+  in mconcat $ replaceHoles 0 holesLocs
+  where
+    showBs :: ByteString -> Text
+    showBs = show @_ @Text . decodeUtf8
+
+    holeLineRegex :: R.Regex
+    holeLineRegex = R.makeRegex ("[ -]+:PLACEHOLDER:[^:]+:" :: Text)
+
+    replacementsMap = Map.fromList allReplacements
+    getReplacement key =
+      Map.lookup key replacementsMap
+      ?: error ("Replacement for key " <> showBs key <> " is not specified")
+
+    pickConfigSubstring :: Int -> Int -> ByteString
+    pickConfigSubstring from len = BS.take len $ BS.drop from rawConfig
+
+    replaceHoles :: Int -> [(R.MatchOffset, R.MatchLength)] -> [ByteString]
+    replaceHoles processedLen [] = one $ BS.drop processedLen rawConfig
+    replaceHoles processedLen ((off, len) : locs) =
+      -- in our case matches here should not overlap
+      assert (off > processedLen) $
+        pickConfigSubstring processedLen (off - processedLen) :
+        replaceHole (pickConfigSubstring off len) ++
+        replaceHoles (off + len) locs
+
+    holeItemRegex :: R.Regex
+    holeItemRegex = R.makeRegex ("(^|:)([ ]*):PLACEHOLDER:([^:]+):" :: Text)
+
+    holeListRegex :: R.Regex
+    holeListRegex = R.makeRegex ("^([ ]*-[ ]*):PLACEHOLDER:([^:]+):" :: Text)
```

Comment:
I supposed that list hole allows inserting a list of entries, for instance [`virtualFiles`](https://github.com/serokell/xrefcheck/pull/57/files#diff-5b93233819d5f91ba783808ea00b7ec4f93d64c58b6b2d1b38e11d692facbd52R34) is now defined as a list placeholder.

Reply:
Yeah, it's clear. I mean that `:PLACEHOLDER:virtualFiles` and, for example, `:PLACEHOLDER:repositoryType` can be matched with `holeItemRegex`, and it seems that the hole always looks like `:PLACEHOLDER:name`, where `name` is just sequence of characters and numbers(?). So that I can't catch why we need 2 different regex and ask for an example.

Reply:
Ah, I see.

Well, in the case of the list, I need to prepend each entry with a `-`. And it is not trivial to figure out automatically where this `-` should be, because yaml has to be properly indented, and indentation may differ depending on the practices used in this particular file. So the simplest I can do is to assume that the placeholder is also prepended with `-` and replicate this `-` at whatever indentation it is.

So the difference is that for lists I expect this `-` before the placeholder, and for single entries the is no such prefix. It turned out that fetching the `-` was simpler with regexps.

Reply:
Oooh, I see now. Thanks!